spec/unit/mutant/ast/regexp_spec.rb

```diff
@@ -0,0 +1,226 @@
+    [/(?!a)/,            :regexp_nlookahead_assertion,     root(r(:nlookahead_assertion, r(:literal_literal, 'a')))],
+    [/(?<!a)/,           :regexp_nlookbehind_assertion,    root(r(:nlookbehind_assertion, r(:literal_literal, 'a')))],
+    [/(?<=a)/,           :regexp_lookbehind_assertion,     root(r(:lookbehind_assertion, r(:literal_literal, 'a')))],
+    [/(?=a)/,            :regexp_lookahead_assertion,      root(r(:lookahead_assertion, r(:literal_literal, 'a')))],
+    [/(?i-:a)+/,         :regexp_options_group,            greedy_options_group],
+    [/(?x-: #{"\n"} )/,  :regexp_options_group,            options_group_with_whitespace],
+    [/foo|bar/,          :regexp_alternation_meta,         alternation],
+    [/foo|bar/,          :regexp_sequence_expression,      alternation]
+  ].freeze
+
+  def indent(string)
+    string.gsub(/^/, ' ' * 6)
+  end
+
+  # rubocop:disable MethodLength
+  # rubocop:disable Metrics/AbcSize
+  def assert_expressions(actual, expected)
+    expect(actual.type).to  be(expected.type)
+    expect(actual.token).to be(expected.token)
+    expect(actual.text).to  eql(expected.text)
+
+    if actual.quantified? != expected.quantified?
+      fail 'Original and generated expressions do not have the same `quantified?` value'
+    end
+
+    if actual.quantified?
+      expect(actual.quantifier.token).to be(expected.quantifier.token)
+      expect(actual.quantifier.text).to eql(expected.quantifier.text)
+      expect(actual.quantifier.mode).to be(expected.quantifier.mode)
+      expect(actual.quantifier.min).to be(expected.quantifier.min)
+      expect(actual.quantifier.max).to be(expected.quantifier.max)
+    end
+
+    if actual.terminal? != expected.terminal?
+      fail 'Original and generated expressions do not have the same `terminal?` value'
+    end
+
+    return if actual.terminal?
+
+    actual.expressions.zip(expected.expressions).each do |(actual_exp, expected_exp)|
+      assert_expressions(actual_exp, expected_exp)
+    end
+  end
+end
+
+RegexpTests::TRANSFORMATIONS.each do |(regex, type, ast)|
+  RSpec.describe Mutant::AST::Regexp::Transformer.lookup(type) do
+    include RegexpTests
+
+    it "transforms #{regex.inspect}" do
+      expression = Regexp::Parser.parse(regex)
+      transform  = Mutant::AST::Regexp.to_ast(expression)
+
+      expect(transform).to eql(ast)
+    end
+
+    next if regex == /\xFF/n
+
+    it "round trips #{regex.inspect}" do
+      original   = Regexp::Parser.parse(regex)
+      ast        = Mutant::AST::Regexp.to_ast(original)
+      expression = Mutant::AST::Regexp.to_expression(ast)
+      round_trip = expression.to_re
+
+      expect_exp_text = indent(YAML.dump(Regexp::Parser.parse(regex)))
+      actual_exp_text = indent(YAML.dump(Mutant::AST::Regexp.to_expression(ast)))
+
+      diff = Mutant::Diff.build(expect_exp_text, actual_exp_text).colorized_diff
+
+      fail_message = <<-MSG
+Failed performing round trip:
```

Comment:
Cool. This part of the spec was also sort of messy and verbose to aid my TDD flow since I had to add 70 new AST types. I plan to clean this file up and make the output a bit less verbose